src/to_vec.jl

```diff
@@ -110,6 +110,34 @@ function to_vec(X::T) where {T<:PermutedDimsArray}
     return x_vec, PermutedDimsArray_from_vec
 end
 
+# Factorizations
+
+function to_vec(x::SVD)
+    x_vec, back = to_vec([x.U, x.S, x.Vt])
+    function SVD_from_vec(v)
+        U, S, Vt = back(v)
+        return SVD(U, S, Vt)
+    end
+    return x_vec, SVD_from_vec
+end
+
+function to_vec(x::Cholesky)
+    x_vec, back = to_vec(x.factors)
+    function Cholesky_from_vec(v)
+        return Cholesky(back(v), x.uplo, x.info)
+    end
+    return x_vec, Cholesky_from_vec
+end
+
+function to_vec(x::S) where {S <: Union{LinearAlgebra.QRCompactWYQ, LinearAlgebra.QRCompactWY}}
+    x_vec, back = to_vec([x.factors, x.T])
+    function QRCompact_from_vec(v)
+        factors, T = back(v)
+        return S(factors, T)
+    end
+    return x_vec, QRCompact_from_vec
+end
+
 # Non-array data structures
```

Comment:
`x.T` consists of upper triangular blocks: https://github.com/JuliaLang/julia/pull/41363#discussion_r659375154
The lower part is undefined on purpose, NaNs can hide there. This line won't work:  
`x_vec, back = to_vec([x.factors, x.T])`

For application purposes it may be preferable to think of derivatives in `Q` and `R`, not in `factors` and `T`.
So `to_vec` would unroll `Q` and `R` into a matrix. To instantiate a new QR object from updated values, one would need to re-calculate the QR factorization:

```
function FiniteDifferences.to_vec(x::S) where {S <: Union{LinearAlgebra.QRCompactWYQ, LinearAlgebra.QRCompactWY}}
    x_vec, x_back = to_vec([x.Q x.R])
    function QRCompact_from_vec(v)
        Q_new, R_new = x_back(v)
        QR = S(Q_new * R_new)
        return S(QR.factors, QR.T)
    end
    return x_vec, QRCompact_from_vec
end
```

Reply:
I tried that but unfortunately recomputing the QR factorization is enough to introduce numerical errors, and make the `to_vec` test fail (since it test for exact equality).

As far as I understand the internal representation used by `to_vec` should not change the result and it should be fine. If it is not the case, I will need to be smarter, somehow.

Reply:
> As far as I understand the internal representation used by to_vec should not change the result and it should be fine. If it is not the case, I will need to be smarter, somehow.

If it contains NaNs, it definitely does. This should be reusing the logic added in JuliaLang/julia#41363, in particular `_triuppers_qr`.

Reply:
I am not sure to understand. The numerical instability I am referring to are those that prevent are introduced by recomputing the QR decomposition internally:

```julia
julia> X = rand(10, 10) ;

julia> F = qr(X) ;

julia> Xnew = F.Q * F.R ;

julia> Xnew == X
false

julia> Fnew = qr(Xnew) ;

julia> Fnew == F
false

julia> Fnew.Q == F.Q
false
```

The tests require `from_vec(to_vec(F)) == F` which is not guaranteed there.

To deal with NaNs in `T` is it preferrable to change the supported version of Julia and use `LinearAlgebra._triuppers_qr` or to reproduce here part of the logic?

Reply:
I just meant that by filtering out the undef values, we get both the correct behavior and avoid any numerical issues related to having to refactorize the recalculated matrix.

> To deal with NaNs in T is it preferrable to change the supported version of Julia and use LinearAlgebra._triuppers_qr or to reproduce here part of the logic?

It's probably easiest if we just copy that code from there for now. Would be good to mention in a comment that this is taken from Base though and that it can be removed once we drop support for older versions of Julia.

Reply:
BTW, a good test here would be to check that `to_vec(qr(A)) == to_vec(qr(A))`

Reply:
I incorporated that, and added a test of NaN in test_to_vec along the way.

QR must unfortunaely be tested separately for now, because it compares the T matrices that have those arbitrary values.